src/deadline_worker_agent/worker.py

```diff
@@ -246,6 +246,10 @@ def run(self) -> None:
                                 grace_time=worker_shutdown.grace_time,
                                 fail_message=worker_shutdown.fail_message,
                             )
+                        else:
+                            # If we are here, it's because self._stop.set() was set causing the monitor_ec2_shutdown thread to join.
+                            # The scheduler thread has a longer wait, so let's wake it up so it can join as well.
+                            self._scheduler._wakeup.set()
```

Comment:
While this does work, calling the private `_wakeup` event doesn't feel ideal. 

It looks like calling `self._scheduler.shutdown()` [also sets the `_wakeup` event](https://github.com/aws-deadline/deadline-cloud-worker-agent/blob/mainline/src/deadline_worker_agent/scheduler/scheduler.py#L1483)

I wonder if we should instead modify [`_monitor_ec2_shutdown`](https://github.com/aws-deadline/deadline-cloud-worker-agent/blob/mainline/src/deadline_worker_agent/worker.py#L364)  to return a [`WorkerShutdown`](https://github.com/aws-deadline/deadline-cloud-worker-agent/blob/mainline/src/deadline_worker_agent/worker.py#L44). This would give the benefit of letting the scheduler handle any edge cases it might need to in the future if the implementation ever changes

Reply:
To me, it doesn't seem like this is the responsibility of the `_monitor_ec2_shutdown()` method. That method should only be responsible for handling EC2-based shutdown events. Instead, it would be ideal if the code that responds to the event:

https://github.com/aws-deadline/deadline-cloud-worker-agent/blob/353d9a25448462484fa8a263318520f63a8015e2/src/deadline_worker_agent/windows/win_service.py#L43-L45

could somehow inform the shutdown grace period and provide the interruption status message when calling `self._scheduler.shutdown(...)`.

Reply:
I considered adding to the monitor_ec2_shutdown function. However, this doesn't just include EC2 shutdowns, it could be a local host. This could also occur when the service is forcefully stopped, but the host is still running. 

I decided to let the scheduler just use the default grace period, rather than define some arbitrary value since we won't have any idea what a good grace period is for this case. We are just making a best effort here to let the worker shut down cleanly.